src/content/language/macros.md

```diff
@@ -0,0 +1,287 @@
+- **No duplicated work** or constant recompiling hurting performance;
+ all the building and code generation happen directly in the compiler,
+ automatically.
+- **Not written to disk**, so no part files or pointers to generated references;
+ macros directly augment the *existing* class.
+- **No confusing/obfuscated testing**;
+ custom diagnostics are emitted like any other message from the analyzer,
+ directly in the IDE.
+
+And also far more efficient, and far less error prone, than manually
+writing solutions to these types of problems yourself.
+
+Check out these examples showing the same JSON serialization
+implemented three different ways:
+
+- Using the [`JsonCodable` macro][].
+- Using the [`json_serializable` code gen package][].
+- Manually, [with `dart:convert`][].
+
+[build_runner]: /tools/build_runner
+[`JsonCodable` macro]: https://github.com/mit-mit/sandbox/blob/main/explorations/json/dart_jsoncodable/bin/main.dart
+[`json_serializable` code gen package]: https://github.com/mit-mit/sandbox/blob/main/explorations/json/dart_json_serializable/bin/main.dart
+[with `dart:convert`]: https://github.com/mit-mit/sandbox/blob/main/explorations/json/dart_convert/bin/main.dart
+
+### Use cases
+
+Macros provide reusable mechanisms to address patterns characterized by tedious
+boilerplate, and often times the need to iterate over the fields of a class.
+Some common examples are:
+
+- **Json serialization.** The extra tooling required to serialize JSON,
+ like the [json_serializable][] package, isn't as efficient as it should be.
+ The `JsonCodable` macro provides a much cleaner way to
+ generate serialization code; [try it today](#the-jsoncodable-macro).
+
+- **Data classes.** Dart's [most requested][] feature is for data classes
+ that automatically provide a constructor, and implementations of the `==`,
+ `hashCode`, and `copyWith()` methods for each field.
+ Implementing the solution with macros would mean users can customize the
+ their data classes however they see fit.
+
+- **Verbose Flutter patterns.** One example is breaking down a complex `build`
+ method into an aggregation of smaller widget classes. It's
+ better for performance and makes the code more maintainable. Unfortunately,
+ writing all those smaller classes requires tons of boilerplate, which discourages
+ users. Macros could potentially provide a solution that iterates over a
+ complex `build` method to generates smaller widget classes,
+ greatly improving productivity and quality of Flutter code.
+
+[json_serializable]: {{site.pub-pkg}}/json_serializable
+[most requested]: {{site.repo.dart.lang}}/issues/314
+
+### How macros work
+
+:::important
+The macros language feature is not stable and currently behind an [experiment flag][].
+Functionality is highly subject to change. This section will remain very high-level until stable.
+:::
+
+To create a macro, you write a macro declaration similar to a class,
```

Comment:
I think we should mention the phases to macro expansion. @jakemac53 can help with that.

Reply:
I am actually not sure if we want to explicitly talk about the phases or not - the API is structured more around "capabilities" (via the different interfaces). Those do correspond to phases, but I am not sure of how much value it would provide to talk about them explicitly?

Reply:
Ah, ok, let's skip that for now

Reply:
Maybe something we could capture as the page/feature matures